### DISCUSSION TITLE
About the visible property of the window

### DISCUSSION THREAD

**Discussion by gqf2008** [marc2332/freya#929] [General]
Delete lines 78 and 108 from freya/crates/renderer/src
/window_state.rs in lines 78 and 108, which has the advantage of not affecting existing functionality and leaving it up to the user to control whether or not the window is displayed, for example the following code

```rs

launch_cfg(
    app, LaunchConfig::LaunchConfig::LaunchConfig
    LaunchConfig::<String>::new()
        .with_size(800., 400.)
        .with_state(state)
        .with_window_attributes(|mut window_attributes| {
            if cfg!(target_os = ‘macos’) {
                window_attributes = window_attributes
                    .with_resizable(false)
                    .with_visible(true)
                    .with_fullsize_content_view(true)
                    .with_title_hidden(true)
                    .with_titlebar_transparent(true);
            }
            window_attributes
        })
        .with_icon(LaunchConfig::load_icon(ICON)), .
)

```

By the way, it would be a good idea to export winit::platform::macos::* and winit::platform::windows::* again, so that the user can have more control over the display of the window

**Comment by marc2332** [comment]
Hey! How is that affecting your app? The window will get visible in L108, so what's the issue?

**Reply by gqf2008** [reply]
Some applications do not need to display a window at startup, but only need to start the system tray, and only need to display a window when the user clicks on the system tray, such as background programs that start up along with the system

**Reply by marc2332** [reply]
Good point, you are right about that 💯

**Comment by marc2332** [comment]
That code was introduced as a workaround for a bug in accesskit, that bug seems like already got fixed a while ago, I tested it myself and it doesn't seem to happen anymore. So I just opened https://github.com/marc2332/freya/pull/930 to remove those lines

**Comment by marc2332** [comment]
> By the way, it would be a good idea to export winit::platform::macos::* and winit::platform::windows::* again, so that the user can have more control over the display of the window

I am not very sold on the idea of reexporting winit window APIs. Ideally, `with_window_attributes` should only be used when you really need it and in that case you just need to add winit as a dep to your project, which I suppose is what you did

**Reply by gqf2008** [reply]
Yes, that's what I'm doing right now, but it may bring trouble from inconsistent versions

**Comment by gqf2008** [comment]
Thanks! I have another request, especially for windows platform, I would like to enable non-main thread to run Eventloop, the application scenario is to wake up the UI interface normally when my application injected into another process via remote thread!

**Reply by marc2332** [reply]
As far I remember running a GUI app in a non-main thread is not really stable outside of Linux. Couldn't you maybe design your app in a way that the main thread is blocked until your other thread notifies it? So basically it would be about swapping what your main and your other thread do

**Reply by gqf2008** [reply]
Yes, running the UI program in a non-main thread can be an option for the user to determine for him/herself. The specific business scenario is: the main thread is the game, and my application is injected as a plugin into the game process to run as a subthread, and my application's UI is only displayed when a hotkey is triggered

**Reply by marc2332** [reply]
Couldn't you keep the freya in a separate process?

**Reply by gqf2008** [reply]
No, because the program is loaded by a dll, not a separate exe.

**Reply by marc2332** [reply]
I'll see if I can add an option to enable running the eventloop in a secondary thread https://github.com/marc2332/freya/issues/943

**Reply by marc2332** [reply]
Just opened https://github.com/marc2332/freya/pull/944

You should now be able to do:

```rs
.with_event_loop_builder(|mut ev| {
    ev.with_any_thread(true);
})
```

**Reply by gqf2008** [reply]
Great!

**Comment by marc2332** [comment]
FYI @gqf2008 I had to revert the removal of the code, but, I added a with_visible method for the launch config. https://github.com/marc2332/freya/pull/935

**Reply by gqf2008** [reply]
That's great.

**Comment by marc2332** [comment]
Thanks for sponsoring me @gqf2008 ! I appreciate it.

**Reply by gqf2008** [reply]
I'm looking forward to the 1.0 release.
